toolchain/sem_ir/clang_decl.h

```diff
@@ -20,17 +20,22 @@ class Decl;
 namespace Carbon::SemIR {
 
 // A Clang declaration mapped to a Carbon instruction.
-// Using custom hashing since the declaration is keyed by the `decl` member for
-// lookup.
-// TODO: Avoid custom hashing by either having the data structure support keying
-// or create a dedicated mapping. See
-// https://discord.com/channels/655572317891461132/768530752592805919/1384999468293947537
+//
+// Note that Clang's AST uses address-identity for nodes, which means the
+// pointer is the canonical way to represent a specific AST node and is expected
+// to be sufficient for comparison, hashing, etc.
 struct ClangDecl : public Printable<ClangDecl> {
   auto Print(llvm::raw_ostream& out) const -> void;
 
-  friend auto CarbonHashtableEq(const ClangDecl& lhs, const ClangDecl& rhs)
-      -> bool {
-    return HashtableEq(lhs.decl, rhs.decl);
+  // Equality comparison uses the address-identity property of the Clang AST and
+  // just compares the `decl` pointers. We canonicalize on these to avoid ever
+  // having different instructions for the same declaration.
```

Comment:
```suggestion
  // Equality comparison uses the address-identity property of the Clang AST and
  // just compares the `decl` pointers. We can disregard `inst_id` because `ClangDecl`s
  // are only used as entries in `SemIR::clang_decls`, which canonicalizes them to
  // ensure that the same `decl` can't be associated with more than one `inst_id`.
```